server/src/main/java/com/cloud/storage/download/managementserver/ManagementServerDownloader.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.cloudstack.engine.subsystem.api.storage.ObjectInDataStoreStateMachine;
+import org.apache.cloudstack.storage.datastore.db.TemplateDataStoreDao;
+import org.apache.cloudstack.storage.datastore.db.TemplateDataStoreVO;
+import org.apache.log4j.Logger;
+
+import com.cloud.storage.ImageStoreDetailsUtil;
+import com.cloud.storage.VMTemplateStoragePoolVO;
+import com.cloud.storage.VMTemplateStorageResourceAssoc;
+import com.cloud.storage.VMTemplateVO;
+import com.cloud.storage.dao.VMTemplateDao;
+import com.cloud.storage.dao.VMTemplatePoolDao;
+import com.cloud.storage.download.managementserver.system.HttpSystemTemplateDownloader;
+import com.cloud.storage.mount.MountManager;
+import com.cloud.template.VirtualMachineTemplate;
+import com.cloud.utils.concurrency.NamedThreadFactory;
+import com.cloud.utils.db.TransactionLegacy;
+import com.cloud.utils.exception.CloudRuntimeException;
+
+public class ManagementServerDownloader {
+    private static final Logger s_logger = Logger.getLogger(ManagementServerDownloader.class);
+
+    @Inject
+    private ImageStoreDetailsUtil imageStoreDetailsUtil;
+
+    @Inject
+    VMTemplatePoolDao vmTemplatePoolDao;
+
+    @Inject
+    private VMTemplateDao templateDao;
+
+    @Inject
+    private TemplateDataStoreDao vmTemplateStoreDao;
+
+    @Inject
+    private MountManager mountManager;
+
+    private static final ExecutorService THREAD_SERVICE = Executors.newCachedThreadPool(new NamedThreadFactory("ManagementServerDownloader"));
+
+    public void download(VirtualMachineTemplate tmpl, DataObject template) {
+        downloadFromUrlToNfs(tmpl, template, template.getUuid());
+    }
+
+
+    public void downloadFromUrlToNfs(VirtualMachineTemplate tmpl, DataObject template, String fileName) {
+        Runnable downloadFile = () -> {
+            Integer nfsVersion = imageStoreDetailsUtil.getNfsVersion(template.getDataStore().getId());
+            String mountPoint = mountManager.getMountPoint(template.getDataStore().getUri(), nfsVersion);
+            VMTemplateVO templateVO = Optional.ofNullable(templateDao.findById(template.getId()))
+                .orElseThrow(() -> new CloudRuntimeException(String.format("Unable to find template by id %d.%n", template.getId())));
+            HttpSystemTemplateDownloader downloader = new HttpSystemTemplateDownloader(tmpl, templateVO, mountPoint);
+            if (downloader.downloadTemplate()) {
+                if (downloader.extractAndInstallDownloadedTemplate()) {
+                    TemplateDownloader.TemplateInformation info = downloader.getTemplateInformation();
+                    try(TransactionLegacy txn = TransactionLegacy.open(ManagementServerDownloader.class.getName())) {
+                        persistTemplate(template, templateVO, info);
+                        persistTemplateStorePoolRef(tmpl, template.getDataStore().getId(), info);
+                        persistTemplateStoreRef(template, info);
+                    }
```

Comment:
no exceptions to be caught? so this is a try once but it doesn't matter if we don't persist action?